srml/evm/src/lib.rs

```diff
@@ -0,0 +1,247 @@
+
+/// Trait that outputs the current transaction gas price.
+pub trait FeeCalculator {
+	/// Return the current gas price.
+	fn gas_price() -> U256;
+}
+
+/// Trait for converting account ids of `balances` module into
+/// `H160` for EVM module.
+///
+/// Accounts and contracts of this module are stored in its own
+/// storage, in an Ethereum-compatible format. In order to communicate
+/// with the rest of Substrate module, we require an one-to-one
+/// mapping of Substrate account to Ethereum address.
+pub trait ConvertAccountId<A> {
+	/// Given a Substrate address, return the corresponding Ethereum address.
+	fn convert_account_id(account_id: A) -> H160;
+}
+
+/// Custom precompiles to be used by EVM engine.
+pub trait Precompiles {
+	/// Try to execute the code address as precompile. If the code address is not
+	/// a precompile or the precompile is not yet available, return `None`.
+	/// Otherwise, calculate the amount of gas needed with given `input` and
+	/// `target_gas`. Return `Ok(Some(status, output, gas_used))` if the execution
+	/// is successful. Otherwise return `Ok(Err(_))`.
+	fn execute(
+		address: H160,
+		input: &[u8],
+		target_gas: Option<usize>
+	) -> Option<core::result::Result<(ExitSucceed, Vec<u8>, usize), ExitError>>;
+}
+
+impl Precompiles for () {
+	fn execute(
+		_address: H160,
+		_input: &[u8],
+		_target_gas: Option<usize>
+	) -> Option<core::result::Result<(ExitSucceed, Vec<u8>, usize), ExitError>> {
+		None
+	}
+}
+
+/// EVM module trait
+pub trait Trait: system::Trait + timestamp::Trait {
+	/// Calculator for current gas price.
+	type FeeCalculator: FeeCalculator;
+	/// Convert account ID to H160;
+	type ConvertAccountId: ConvertAccountId<Self::AccountId>;
+	/// Currency type for deposit and withdraw.
+	type Currency: Currency<Self::AccountId>;
+	/// The overarching event type.
+	type Event: From<Event> + Into<<Self as system::Trait>::Event>;
+	/// Precompiles associated with this EVM engine.
+	type Precompiles: Precompiles;
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Example {
+		Accounts get(fn accounts) config(): map H160 => Account;
+		AccountStorages: double_map H160, blake2_256(H256) => H256;
+	}
+}
+
+decl_event!(
+	/// EVM events
+	pub enum Event {
+		/// Ethereum events from contracts.
+		Log(Log),
+	}
+);
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn deposit_event() = default;
+
+		#[weight = SimpleDispatchInfo::FixedNormal(10_000)]
+		fn deposit_balance(origin, value: BalanceOf<T>) -> Result {
+			let sender = ensure_signed(origin)?;
+
+			T::Currency::withdraw(
+				&sender,
+				value,
+				WithdrawReason::Reserve,
+				ExistenceRequirement::KeepAlive,
+			)?;
+
+			let bvalue = U256::from(UniqueSaturatedInto::<u64>::unique_saturated_into(value));
+			let address = T::ConvertAccountId::convert_account_id(sender);
+			Accounts::mutate(&address, |account| {
+				account.balance += bvalue;
```

Comment:
will it be more performant if the account balance is moved to its own storage? otherwise this will unnecessarily read and write `Account::code` which can be large.

Reply:
I moved code to its own storage. I'm currently still keeping `balance` and `nonce` in one place because they're usually changed together.